### PR TITLE
Redesign

### PR DESCRIPTION
**Pull request by j-m**
Refactored existing code and then optimised (also looked at the generated js to optimise).
Made constructor args optional.
Added getters.
Added benchmark test.
Removed _ from variable names.
Added semicolons.
Changed tsconfig.
Updated package.json.

### CODE REVIEW

**Comment by jmsv** [comment]
Changing code style and tsconfig was unnecessary - please revert these so meaningful changes are more readable

**Comment by j-m** [comment]
Why should `noImplicitAny` be `false`?

File: index.ts [hunk 1]
```diff
@@ -1,35 +1,65 @@
...
 
   get length(): number {
-    return this._original.length
+    return this.original.length
   }
 
-  constructor(input: object[], keys: string[]) {
-    this._original = input.map((x) => Object.assign({}, x))
-    this._keys = keys
+  public getKeys(): string[] | undefined {
```
Comment (jmsv) [inline index.ts]: This isn't how getters/setters work - please see https://www.typescriptlang.org/docs/handbook/classes.html

should be something like:
```typescript
get keys(): string[] {
  return this.keys
}
```

File: index.ts [hunk 2]
```diff
@@ -1,35 +1,65 @@
...
+    return this.keys
+  }
 
-    this._transposed = transpose(this._keys, this._original)
+  public getOriginal(): object[] {
```
Comment (jmsv) [inline index.ts]: should be a getter function - see above comment

File: index.ts [hunk 3]
```diff
@@ -1,35 +1,65 @@
...
+    return this.original
   }
 
-  public get(keyName: string, keyValue: string): object[] {
-    return this._transposed[keyName][keyValue]
+  public getTransposed(): object[] {
```
Comment (jmsv) [inline index.ts]: getter

File: index.ts [hunk 4]
```diff
@@ -1,35 +1,65 @@
...
 
-const transpose = (keys: string[], items: object[], current?: object) => {
-  return keys.reduce((transposed, key) => {
-    transposed[key] = items.reduce((acc, item) => {
-      acc[item[key]] ? acc[item[key]].push(item) : acc[item[key]] = [item]
-      return acc
-    }, current || {})
-    return transposed
-  }, {})
+  public push(...data: object[]) {
```
Comment (jmsv) [inline index.ts:40]: parameter name of `items` was more descriptive imo

File: index.ts [hunk 5]
```diff
@@ -1,35 +1,65 @@
...
+    data.forEach((datum) => {
+      const clone = Object.assign({}, datum)
+      this.original.push(clone)
+      this.transpose(clone)
+    })
+    return this.length
+  }
+
+  private getKeysForObject(datum: object): string[] {
```
Comment (jmsv) [inline index.ts]: not sure why we need this

File: index.ts [hunk 6]
```diff
@@ -1,35 +1,65 @@
...
+    return this.keys !== undefined ? this.keys : Object.keys(datum)
+  }
+
+  private transpose(datum: { [key: string]: any }) {
+    this.getKeysForObject(datum).forEach((key) => {
```
Comment (jmsv) [inline index.ts]: why does the `transpose` use a `forEach` rather that a `reduce` now? I feel like this might have a performance impact

File: index.ts [hunk 7]
```diff
@@ -1,35 +1,65 @@
...
+      if (datum[key] === undefined) {
+        return
+      }
+      if (this.transposed[key] === undefined) {
```
Comment (jmsv) [inline index.ts]: lots of `undefined` checks in this method that weren't needed before - why now?

File: package.json [hunk 8]
```diff
@@ -18,13 +18,22 @@
     "url": "git+https://github.com/jmsv/oaty.git"
   },
   "keywords": [
+    "array",
+    "keys",
     "oaty",
     "object",
-    "array",
```
Comment (jmsv) [inline package.json:23]: why rm 'array'? 🤔

File: test/benchmark.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+}, {
+  a: 1,
+  b: 3,
+  fruit: 'banana'
+}]
+
+var results = []
+
+function time(runs) {
+    var oaty = new OatyArray({data: testArray});
```
Comment (jmsv) [inline test/benchmark.js]: no code style enforced for test stuff atm but opting to not use semicolons - think [standardjs](https://github.com/standard)

File: index.ts [hunk 10]
```diff
@@ -1,35 +1,65 @@
 export class OatyArray {
-  private _original: object[]
-  private _keys: string[]
-  private _transposed: object = {}
+  private keys: string[] | undefined
+  private original: object[] = []
+  private transposed: { [key: string]: any } = {}
+
+  constructor(options?: { data?: object[], keys?: string[] }) {
```
Comment (jmsv) [inline index.ts]: I'm not too keen on `OatyArray` only taking an options object as a parameter. Would prefer the initial array and an `options` object, e.g.

```ts
constructor(input?: array, options?: { keys?: string[] }) {
```

so you can init a new instance as, for example:
```ts
const empty = new OatyArray()
const emptyWithKeys = new OatyArray([], { keys: ['a', 'b'] })
const arrayNoKeys = new OatyArray([...])
```

File: index.ts [hunk 11]
```diff
@@ -1,35 +1,65 @@
...
-    return this._transposed[keyName][keyValue]
+  public getTransposed(): object[] {
+    return this.original
   }
 
-  public push(...items: object[]): number {
-    this._transposed = transpose(this._keys, items, this._transposed)
-    return this._original.push(...items)
+  public get(keyName: string, keyValue: string): object[] | undefined {
+    if (this.transposed === undefined
```
Comment (jmsv) [inline index.ts]: loads of undefined checks?

File: test/benchmark.js [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+  
+  function getNanoseconds(hrtime) {
+    return hrtime[0] * 1e9 + hrtime[1];
+  }
+  
+  function getAverage(values) {
+    return values.reduce((previous, current) => current += previous) / values.length;
+  }
+  
+  time(1)
```
Comment (jmsv) [inline test/benchmark.js]: probably better to use a benchmarking/timing library for this and randomised, variable-length datasets

**Review by jmsv — CHANGES_REQUESTED** [review]
tl;dr:
- use proper get/setters rather than java-esque `getThing()` functions.
- need to sort out better constructor params.
- `transpose`/`get` functions now have loads of `=== undefined` checks and use `forEach` rather than `reduce`, without explanation. if this is for a reason we should discuss it, otherwise revert.

File: index.ts (hunk 1 above)
Comment (j-m) [reply index.ts]: 🤯

File: index.ts (hunk 5 above)
Comment (j-m) [reply index.ts]: If options.keys is defined then we want to transpose only those keys, otherwise we'll use the object's.

File: index.ts (hunk 6 above)
Comment (j-m) [reply index.ts]: Read somewhere that reduce, filter, map, and forEach all have roughly the same performance but that the others transform the data, imo using reduce was somewhat misleading.

File: index.ts (hunk 11 above)
Comment (j-m) [reply index.ts]: Same as another comment: If options.keys is defined then we want to transpose only those keys, otherwise we'll use the object's.

But we do need to decide if we want to return something even if it's just an empty array; would make it more consistent and look like it actually tried to search.

File: index.ts (hunk 6 above)
Comment (jmsv) [reply index.ts]: the `reduce` was transforming the array down to a single value: the transformed object - this is exactly the kind of thing reduce is for. would rather this change is reverted

File: index.ts (hunk 4 above)
Comment (j-m) [reply index.ts:40]: I wanted to standardise the name for it; you used `input` in the constructor.

File: index.ts (hunk 7 above)
Comment (j-m) [reply index.ts]: I wanted to just do `this.transposed[key][datum[key]] = [datum]` but it throws if `this.transposed[key]` is undefined. 
Same as another comment: If options.keys is defined then we want to transpose only those keys, otherwise we'll use the object's.

File: package.json (hunk 8 above)
Comment (j-m) [reply package.json:23]: I didn't, I alphabetised.

File: test/benchmark.js (hunk 9 above)
Comment (j-m) [reply test/benchmark.js]: I read into ASI and you have to be more careful with JS. I don't think it's worth the risk and that semicolons make it abundantly clear.

File: test/benchmark.js (hunk 12 above)
Comment (j-m) [reply test/benchmark.js]: Yea, just wanted something basic for the mean time.
After this is merged we could create a `profiling` branch and make commit all future integration or memory/time profiling tests against that.

File: tsconfig.json [hunk 13]
```diff
@@ -5,7 +5,7 @@
     "declaration": true,
     "outDir": "./dist",
     "strict": true,
-    "noImplicitAny": false,
+    "noImplicitAny": true,
```
Comment (j-m) [inline tsconfig.json:8]: I feel like having implicit `any`s defeats the point of TS

File: index.ts (hunk 5 above)
Comment (jmsv) [reply index.ts]: do we reuse this code enough to justify a new function though? 

also, the body of this function does:
```ts
this.keys !== undefined ? this.keys : Object.keys(datum)
```

which is equivalent to:
```ts
this.keys || Object.keys(datum)
```

which is shorter and much more readable

File: tsconfig.json (hunk 13 above)
Comment (jmsv) [reply tsconfig.json:8]: No, just makes types optional. I just made this `false` when I was prototyping but probably a good idea to make it true now

File: index.ts (hunk 10 above)
Comment (j-m) [reply index.ts]: As it stands you can omit or redorder the args, like this:
`new OatyArray()`
`new OatyArray({})`
`new OatyArray({ keys: ['a', 'b'] })`
`new OatyArray({ data: [...] })`
`new OatyArray({ data: [...], keys: ['a', 'b'] })`
`new OatyArray({ keys: ['a', 'b'], data: [...] })`

This is more future proof and keeps it customisable for when we add more options.

File: index.ts [hunk 14]
```diff
@@ -1,35 +1,65 @@
...
-  public get(keyName: string, keyValue: string): object[] {
-    return this._transposed[keyName][keyValue]
+  public getTransposed(): object[] {
+    return this.original
   }
 
-  public push(...items: object[]): number {
-    this._transposed = transpose(this._keys, items, this._transposed)
-    return this._original.push(...items)
+  public get(keyName: string, keyValue: string): object[] | undefined {
```
Comment (j-m) [inline index.ts]: Also, should we rename this to `find` to be more consistent with other things?

File: index.ts (hunk 11 above)
Comment (jmsv) [reply index.ts]: not sure why `.get` needed the rewrite? imo less readable now

---

> make it look like it actually tried to search

I feel like this is just misleading the consumer of the library - would prefer we returned `undefined`, matching the behaviour of `Array.prototype.find`

File: index.ts (hunk 5 above)
Comment (j-m) [reply index.ts]: Man have I missed JS. 
Absolutely changing!

File: index.ts (hunk 1 above)
Comment (j-m) [reply index.ts]: Ahhhh now i understand why you need the underscores, silly JS

File: index.ts (hunk 10 above)
Comment (jmsv) [reply index.ts]: I understand that, but think the `options` parameter should just be for options, rather than options _and_ data.

File: index.ts (hunk 14 above)
Comment (jmsv) [reply index.ts]: Might be misleading - `Array` `find` takes a function parameter, `OatyArray` `get` doesn't

File: test/benchmark.js (hunk 9 above)
Comment (jmsv) [reply test/benchmark.js]: I've been writing TS/JS without semicolons for a while and haven't had any problems. I'm standing by this 😁

File: package.json (hunk 8 above)
Comment (jmsv) [reply package.json:23]: oh yeah, y tho

File: index.ts (hunk 4 above)
Comment (jmsv) [reply index.ts:40]: I felt like `input` was more descriptive for `constructor`, and `items` was more descriptive for `push`, but agree it might be a good idea to make them uniform. '`data`' seems generic though

File: index.ts (hunk 7 above)
Comment (jmsv) [reply index.ts]: imo, `=== undefined` is rarely needed, but this relatively short function now uses it 3 times

File: index.ts (hunk 1 above)
Comment (jmsv) [reply index.ts]: :joy:

File: index.ts (hunk 10 above)
Comment (j-m) [reply index.ts]: Ohhh I'm with you now, agreed.

File: index.ts (hunk 4 above)
Comment (j-m) [reply index.ts:40]: I've not tested it but my intention is to be able to push objects of different classes, so having a generic name is probably better.

File: test/benchmark.js (hunk 9 above)
Comment (j-m) [reply test/benchmark.js]: Okay, I trust you

File: index.ts (hunk 11 above)
Comment (j-m) [reply index.ts]: `.get` needs the undefined checks.
Surely `return this._transposed[keyName][keyValue]` would've thrown an NPE if nothing had been pushed or if that key hadn't been transposed?

Comment (j-m) [reply index.ts]: `find` also returns a singular object for when there's 1 result rather than an array of 1 object. 
For me at least, consistently returning an array is easier to handle.

File: index.ts (hunk 6 above)
Comment (j-m) [reply index.ts]: Right but you had to use two `reduces` to get the same effect as one `forEach`

File: index.ts (hunk 11 above)
Comment (j-m) [reply index.ts]: Oh wait I forgot `_this.transposed` is initialised to {} . I'll remove the first undefined check, but I'll have to add one depending on what we return

Comment (jmsv) [reply index.ts]: > `find` also returns a singular object for when there's 1 result rather than an array of 1 object.

that's because it searches for the first match, rather than all matches

Comment (j-m) [reply index.ts]: oh really? well, that sucks. Soooo, am I returning undefined or [], or is this a new option?

Comment (jmsv) [reply index.ts]: return matching object array directly (effectively return `undefined` if there's no match)

Comment (jmsv) [reply index.ts]: if we added a function similar to `Array.prototype.filter` in the future, that'd return `[]` if no matches. Our `get` function is more like Array's `find` so returns `undefined`

File: index.ts (hunk 7 above)
Comment (j-m) [reply index.ts]: Is there a better way to `this.transposed[key][datum[key]].push(datum)` given `this.transposed[key]` or `this.transposed[key][datum[key]]` could be undefined?

File: index.ts (hunk 11 above)
Comment (j-m) [reply index.ts]: Hmmm, would it not make more sense to return `[]` for consistency? Our `get` and `Array.find` are very different, and given we're not calling it `find` I don't think there's any risk of someone expecting `undefined`

Comment (jmsv) [reply index.ts]: If I was querying data that doesn't exist, `undefined` is exactly what I'd expect

File: index.ts (hunk 4 above)
Comment (jmsv) [reply index.ts:40]: but in its current state it only accepts `object[]` - sort of makes more sense to rename when it needs a more accurate name, rather than in anticipation of a feature

File: index.ts (hunk 11 above)
Comment (j-m) [reply index.ts]: Okay, I'll leave it as undefined. I will make it configurable ([], undefined, or throw) in another branch.

File: index.ts [hunk 15]
```diff
@@ -1,35 +1,60 @@
...
-  private _keys: string[]
-  private _transposed: object = {}
+  private _keys: string[] | undefined
+  private _original: object[] = []
+  private _transposed: { [key: string]: any } = {}
+
+  constructor(data?: object[], options?: { keys?: string[] }) {
+    if (data !== undefined) { this.push(...data) }
+    if (options !== undefined) {
+      if (options.keys !== undefined) { this._keys = options.keys }
```
Comment (j-m) [inline index.ts]: Just realised I can safely remove this undefined check (`options.keys !== undefined`)

File: index.ts (hunk 4 above)
Comment (j-m) [reply index.ts:40]: I couldn't think of anything to rename it to where it made sense in both contexts so I left it as `data`.

File: index.ts (hunk 1 above)
Comment (j-m) [reply index.ts]: Resolved

File: index.ts (hunk 2 above)
Comment (j-m) [reply index.ts]: Resolved

File: index.ts (hunk 3 above)
Comment (j-m) [reply index.ts]: Resolved

File: index.ts (hunk 5 above)
Comment (j-m) [reply index.ts]: Resolved

File: test/benchmark.js (hunk 9 above)
Comment (j-m) [reply test/benchmark.js]: Resolved

File: index.ts (hunk 10 above)
Comment (j-m) [reply index.ts]: Resolved

File: index.ts (hunk 11 above)
Comment (j-m) [reply index.ts]: Resolved. Will do as another feature.

File: index.ts (hunk 15 above)
Comment (jmsv) [reply index.ts]: Go for it :100:

File: index.ts [hunk 16]
```diff
@@ -1,35 +1,60 @@
 export class OatyArray {
-  private _original: object[]
-  private _keys: string[]
-  private _transposed: object = {}
+  private _keys: string[] | undefined
+  private _original: object[] = []
+  private _transposed: { [key: string]: any } = {}
+
+  constructor(data?: object[], options?: { keys?: string[] }) {
+    if (data !== undefined) { this.push(...data) }
```
Comment (jmsv) [inline index.ts]: these `undefined` checks are still unnecessary and should be rewritten imo, for example, this line can just be:

```ts
this.push(...data || [])
```

File: index.ts (hunk 15 above)
Comment (jmsv) [reply index.ts]: just realised, you have an `if` in an `if`, and neither are needed:

```ts
if (options !== undefined) {
  if (options.keys !== undefined) { this._keys = options.keys }
}
```

change this to
```ts
this._keys = options && options.keys || []
```

File: index.ts [hunk 17]
```diff
@@ -1,35 +1,60 @@
...
+    return this._original
+  }
 
-    this._transposed = transpose(this._keys, this._original)
+  get transposed(): object[] {
+    return this._original
   }
 
-  public get(keyName: string, keyValue: string): object[] {
+  public get(keyName: string, keyValue: string): object[] | undefined {
```
Comment (jmsv) [inline index.ts]: this whole function can just be

```ts
return this._transposed[keyName] ? this._transposed[keyName][keyValue] : undefined
```

File: index.ts [hunk 18]
```diff
@@ -1,35 +1,60 @@
...
 
-const transpose = (keys: string[], items: object[], current?: object) => {
-  return keys.reduce((transposed, key) => {
-    transposed[key] = items.reduce((acc, item) => {
-      acc[item[key]] ? acc[item[key]].push(item) : acc[item[key]] = [item]
-      return acc
-    }, current || {})
-    return transposed
-  }, {})
+  private transpose(datum: { [key: string]: any }) {
```
Comment (jmsv) [inline index.ts]: I don't understand the rewrite here and prefer the original; it's more concise and imo more readable. Think `reduce` might also be higher performance than `forEach` used in this rewrite

File: test/test.js [hunk 19]
```diff
@@ -16,44 +16,54 @@ const testArray = [{
...
-      { a: 6, b: 6, fruit: 'courgette' }
-    )
-
-    const matches = oat.get('fruit', 'potato')
-
-    expect(matches).to.not.be.undefined
-    expect(matches[0]).to.not.be.undefined
-    expect(matches[0].b).to.equal(5)
+  describe('Push', () => {
+    it('push function should return count', () => {
```
Comment (jmsv) [inline test/test.js]: usually `describe` and `it` are meant to make sense in terms of grammar - maybe make it something like:

```ts
describe('push function', () => {
  it('should return count', () => {
```

which reads as "push function should return count", for example

**Review by jmsv — CHANGES_REQUESTED** [review]

File: index.ts (hunk 15 above)
Comment (j-m) [reply index.ts]: Man I love JS.
Will change it to 
`this._keys = options.keys || []`
as we'd be duplicating `options &&` when we add more options

File: test/test.js (hunk 19 above)
Comment (j-m) [reply test/test.js]: I was going to rename them all when we add more tests to follow "given when then" but I'll do it now

File: index.ts (hunk 17 above)
Comment (j-m) [reply index.ts]: Was more in preparation for the future when we'll add different returns (undefined, [], throw) for when an object isn't found.

File: index.ts (hunk 16 above)
Comment (j-m) [reply index.ts]: Then I may as well use defaults and do
`constructor(data: object[] = [], options?: { keys?: string[] }) {
    this.push(...data)`

File: index.ts (hunk 15 above)
Comment (j-m) [reply index.ts]: I'm going to use 
`if (options !== undefined) {
  this._keys = options.keys
}`
Then we won't have to do `options &&` for each new option we have,  and I actually rely on `_keys` being undefined rather than empty for `(this._keys || Object.keys(datum))` in `transpose`.

File: index.ts (hunk 16 above)
Comment (j-m) [reply index.ts]: Imo `if (data !== undefined)` is more readable than having a default

File: index.ts (hunk 17 above)
Comment (jmsv) [reply index.ts]: "Premature optimization is the root of all evil" - Donald Knuth

File: index.ts (hunk 16 above)
Comment (jmsv) [reply index.ts]: I think default params to keep it concise

File: index.ts (hunk 15 above)
Comment (jmsv) [reply index.ts]: Then stop relying on `_keys` being `undefined` - Was this a problem with the original `transpose` implementation?

File: test/test.js (hunk 19 above)
Comment (j-m) [reply test/test.js]: In fact, I also need to add some more tests so I'll remove the data and use fixtures

File: index.ts (hunk 17 above)
Comment (j-m) [reply index.ts]: I agree but it'll be the next thing I do after everything on this review as it's something I need

File: index.ts (hunk 15 above)
Comment (j-m) [reply index.ts]: Yes. The new transpose is fundamentally different: I'm adding keys based on the object unless `_keys` is defined, in which case I'm only transposing those.

File: index.ts (hunk 16 above)
Comment (j-m) [reply index.ts]: Okay

File: index.ts (hunk 6 above)
Comment (j-m) [reply index.ts]: Continued in next review

File: index.ts (hunk 7 above)
Comment (j-m) [reply index.ts]: Continued in next review

File: index.ts (hunk 15 above)
Comment (jmsv) [reply index.ts]: So it sounds like the only difference is how it selects keys?

File: index.ts [hunk 20]
```diff
@@ -1,35 +1,60 @@
 export class OatyArray {
-  private _original: object[]
-  private _keys: string[]
-  private _transposed: object = {}
+  private _keys: string[] | undefined
+  private _original: object[] = []
+  private _transposed: { [key: string]: any } = {}
+
+  constructor(data?: object[], options?: { keys?: string[] }) {
```
Comment (j-m) [inline index.ts]: I think I may revert back to having `data` inside `options`.
I don't really want to have to specify initialising with an empty array, and I can't do 
`OatyArray({keys: [...]})` because it thinks I'm trying to set data to `{keys:[...]}` which doesn't work

File: index.ts (hunk 18 above)
Comment (j-m) [reply index.ts]: reduce is nowhere near as readable; I still have no clue how it work. I don't see how what is essentially is a nested for, could have better performance than just one

File: index.ts (hunk 20 above)
Comment (j-m) [reply index.ts]: Reverted

File: test/test.js (hunk 19 above)
Comment (j-m) [reply test/test.js]: Outdated

File: index.ts (hunk 18 above)
Comment (j-m) [reply index.ts]: Outdated

File: index.ts (hunk 17 above)
Comment (j-m) [reply index.ts]: Outdated

File: index.ts (hunk 16 above)
Comment (j-m) [reply index.ts]: Outdated. Data is an optional parameter again

File: index.ts [hunk 21]
```diff
@@ -1,35 +1,52 @@
 export class OatyArray {
-  private _original: object[]
-  private _keys: string[]
-  private _transposed: object = {}
+  private _keys: string[] | undefined
+  private _original: object[] = []
+  private _transposed: { [key: string]: any } = {}
+
+  constructor(options?: { data?: object[], keys?: string[] }) {
```
Comment (j-m) [inline index.ts]: `data` is optional again

File: index.ts [hunk 22]
```diff
@@ -1,35 +1,52 @@
...
+  }
 
-    this._transposed = transpose(this._keys, this._original)
+  get transposed(): object[] {
+    return this._original
   }
 
   public get(keyName: string, keyValue: string): object[] {
-    return this._transposed[keyName][keyValue]
+    return this._transposed[keyName][keyValue] || []
```
Comment (j-m) [inline index.ts]: We can make this customisable. For me, it is more useful to return an empty array when nothing is found.

File: index.ts [hunk 23]
```diff
@@ -1,35 +1,52 @@
...
+  get original(): object[] {
+    return this._original
+  }
 
-    this._transposed = transpose(this._keys, this._original)
+  get transposed(): object[] {
+    return this._original
   }
 
   public get(keyName: string, keyValue: string): object[] {
```
Comment (j-m) [inline index.ts]: Given not all data has to have all keys, it may be useful to make `keyValue` optional so that I could get all objects that have that key in O(1), too.

Comment (j-m) [inline index.ts]: Given not all data has to have all keys, it may be useful to make keyValue optional so that I could get all objects that have that key in O(1), too.

File: index.ts [hunk 24]
```diff
@@ -1,35 +1,52 @@
...
-    return this._transposed[keyName][keyValue]
+    return this._transposed[keyName][keyValue] || []
   }
 
-  public push(...items: object[]): number {
-    this._transposed = transpose(this._keys, items, this._transposed)
-    return this._original.push(...items)
+  public push(data: object[]) {
```
Comment (j-m) [inline index.ts]: Rest/spread parameters are still new to me; I didn't mean to commit this. 
Are the before and after equivilent?

File: package.json [hunk 25]
```diff
@@ -5,8 +5,10 @@
   "main": "dist/index.js",
   "types": "dist/index.d.ts",
   "scripts": {
-    "build": "tsc",
-    "test": "npm run build && mocha --reporter spec",
+    "clean": "rimraf dist",
+    "benchmark": "npm run build && node dist/test/benchmark.js",
+    "build": "npm run clean && tsc",
+    "test": "mocha --require ts-node/register test/**/*.spec.ts --reporter list",
```
Comment (j-m) [inline package.json:11]: Changed the reporter to `list` as `spec` took up too much space imo

Comment (jmsv) [reply package.json:11]: cool

File: index.ts (hunk 22 above)
Comment (jmsv) [reply index.ts]: alright

File: index.ts (hunk 24 above)
Comment (jmsv) [reply index.ts]: `push`ing without spreading is different, e.g. given:

```ts
const values = [1, 2, 3]
const array = [0]

array.push(values) // [0, [1, 2, 3]]
// or
array.push(...values) // [0, 1, 2, 3]
```

File: index.ts (hunk 21 above)
Comment (jmsv) [reply index.ts]: can you move `data` back outside of `options`?

File: dist/test/benchmark.js [hunk 26]
```diff
@@ -0,0 +1,42 @@
+"use strict";
```
Comment (jmsv) [inline dist/test/benchmark.js]: why are the tests ending up in the `dist` directory?

File: index.ts (hunk 21 above)
Comment (j-m) [reply index.ts]: Not really, if I wanted to set just the `keys` I would rather call `new OatyArray({keys: ["a"]})` than `new OatyArray([], {keys: ["a"]})`.

File: index.ts (hunk 24 above)
Comment (j-m) [reply index.ts]: Ah okay, makes sense. Will revert. Spreading is clearly the more desirable option here

File: index.ts (hunk 15 above)
Comment (j-m) [reply index.ts]: Whoops, missed this. Yes, which is why the new version looked had more `if`s than before. 
Outdated.

File: dist/test/benchmark.js (hunk 26 above)
Comment (j-m) [reply dist/test/benchmark.js]: Whoops, didn't commit the .tsconfig `exclude` change

Comment (j-m) [reply dist/test/benchmark.js]: I had excluded `"**/*.spec.ts"` although this won't exclude `test/benchmark.ts`. I'm not entirely sure where we should put Volumetric & Performance stuff. Thoughts?

File: index.ts (hunk 21 above)
Comment (jmsv) [reply index.ts]: In most cases an `OatyArray` would be initialised with data - I prefer `new OatyArray(data, {})`

File: index.ts (hunk 24 above)
Comment (jmsv) [reply index.ts]: yeah `push`ing an array just doesn't make sense

File: dist/test/benchmark.js (hunk 26 above)
Comment (jmsv) [reply dist/test/benchmark.js]: new directory called `benchmarks` or something?

File: index.ts (hunk 21 above)
Comment (j-m) [reply index.ts]: We should discourage initialising with data, though. 
In my case, I am generating the data. It is much faster to add data when it's created than to add it all at the end. You should only pass in the data if comes from a third party source.

Comment (jmsv) [reply index.ts]: > We should discourage initialising with data, though.

Disagree with this - your case isn't everyone's case

Comment (j-m) [reply index.ts]: It's not about that. 
`data` cannot be optional if it is outside of `options`.
Passing `null`, `void`, `[]`, `undefined` etc for a mandatory field is bad design; this is the exact purpose of overloads and default/optional arguments.

Comment (jmsv) [reply index.ts]: I don't think it's bad design. You initialise an empty array by doing `let arr = []` and I think it makes sense for `OatyArray` to take `[]` as data for empty initialisation.

Regardless, `data` is not an 'option'; the `options` param should be reserved for… options

Comment (j-m) [reply index.ts]: I don't agree but I will revert and we can discuss this another time

File: index.ts [hunk 27]
```diff
@@ -1,35 +1,58 @@
...
 
-const transpose = (keys: string[], items: object[], current?: object) => {
-  return keys.reduce((transposed, key) => {
-    transposed[key] = items.reduce((acc, item) => {
-      acc[item[key]] ? acc[item[key]].push(item) : acc[item[key]] = [item]
-      return acc
-    }, current || {})
-    return transposed
-  }, {})
+  private transpose(data: object[]) {
```
Comment (jmsv) [inline index.ts:45]: I still prefer the reduce, as before

**Review by jmsv — APPROVED** [review]
looks good :100:

please add an overview of the changes in the readme changlelog under `0.3.0`, and we can merge, bump the version and publish
